pennylane/ops/functions/is_unitary.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the qml.is_unitary function.
+"""
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.operation import Operator
+
+
+def is_unitary(op: Operator):
+    """Check if the operation is unitary."""
+    return np.allclose(
+        qml.prod(op, qml.adjoint(op)).matrix(), np.identity(2**op.num_wires)
+    ) and np.allclose(qml.prod(qml.adjoint(op), op).matrix(), np.identity(2**op.num_wires))
```

Comment:
I think we could break up this with some explanatory variables instead of one line that covers three lines.  Also, `qml.math.allclose` would cover comparison in the various interfaces, like torch and tensorflow.

Yes, this takes a bunch more lines now, but its not so smushed up. We also don't have to create the identity matrix and adjoint op multiple times.

```suggestion
    identity_mat = np.identity(2**op.num_wires)
    adj_op = qml.adjoint(op)
    
    op_prod_adjoint_matrix = qml.prod(op, adj_op).matrix()
    if not qml.math.allclose(op_prod_adjoint_matrix, identity_mat):
        return False
    adj_prod_op_matrix = qml.prod(adj_op, op).matrix()
    return qml.math.allclose(adj_prod_op_matrix , identity_mat)
```

Reply:
Yep, you are right:

> *OO† = I*
> *O<sup>-1</sup>OO† = O<sup>-1</sup>*
> *O† = O<sup>-1</sup>*

Reply:
The above would only work if *O* is invertible, but we already know it is because *OO† = I*